python/sensr_message_listener.py

```diff
@@ -39,78 +45,114 @@ def __init__(self,
         self._listener_type = listener_type
         self._output_ws = None
         self._point_ws = None
+        self._state = MessageListener.State.STOPPED
         
 
     def is_output_message_listening(self):
         return self._listener_type == ListenerType.OUTPUT_MESSAGE or self._listener_type == ListenerType.BOTH
 
     def is_point_result_listening(self):
         return self._listener_type == ListenerType.POINT_RESULT or self._listener_type == ListenerType.BOTH
+    
+    def check_oveflow_error(self, output):
+        if output.HasField('event') and output.event.HasField('health'):
+            if output.event.health.master == SystemHealth.Status.OUTPUT_BUFFER_OVERFLOW:
+                return True
+        return False
 
     async def _output_stream(self):
-        async with websockets.connect(self._output_address, ssl=self._ssl_context, compression=None, max_size=None, ping_interval=None) as websocket:
-            self._output_ws = websocket
-            while self._is_running:
-                try:
-                    message = await asyncio.wait_for(websocket.recv(), timeout=1.0) # Receive output messages from SENSR                    
-                    output = OutputMessage()
-                    output.ParseFromString(message)
-                    self._on_get_output_message(output)
-                except asyncio.TimeoutError:
-                    pass
-            # Clean up websocket connection
-            await self._output_ws.close()
-            # if point websocket is already closed, then call loop.stop()
-            # loop.stop should be called to exit `loop.run_forever`
-            if ((self._point_ws == None) or 
-                (self._point_ws != None and self._point_ws.closed == True)):
-                if (self._loop != None and self._loop.is_running()):
-                    self._loop.stop()
+       while self._state != MessageListener.State.STOPPED and self._state != MessageListener.State.STOP_REQUESTED:
```

Comment:
Can you explain a bit why this while-statement is here? 
I thought `_output_stream` was already running in a loop somewhere based on the code `self._loop.create_task(self._output_stream())` below.

Reply:
Ah yes, the codes in python is quite difficult to understand. 
At the first implementation, I was trying to create a new task when reconnect is requested.
But creating a new task while event loop is running was not possible.
So, I tweaked `_output_stream` to run forever until listener's state is STOPPED.
Rough sequence of reconnect is below.

1. a client code call reconnect().
2. `_main()` in sensr_message_listener close the connection to SENSR.
3. `_output_stream()` will exit `while not self._output_ws.closed:` loop because the connection is closed.
4. `_output_stream()` will call `websockets.connection` because of the loop you asked.
5. `_state` will be changed to RUNNING by `_main()`